WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+	// Apple's original implementation cannot find appropriate dividers for a higher pixel clock
+	// value like 533.25 MHz (HDMI 2.0 4K @ 60Hz) and then causes an infinite loop in the kernel.
+	//
+	// This reverse engineering research focuses on identifying fields in CRTC parameters by
+	// carefully analyzing Apple's original implementation, and a better implementation that
+	// conforms to Intel Graphics Programmer Reference Manual is provided to fix the issue.
+	//
+	// This is the first stage to try to solve the HDMI 2.0 output issue on Dell XPS 15 9570,
+	// and is now succeeded by the LSPCON driver solution.
+	// LSPCON is used to convert DisplayPort signal to HDMI 2.0 signal.
+	// When the onboard LSPCON chip is running in LS mode, macOS recognizes the HDMI port as
+	// a HDMI port. Consequently, ComputeHdmiP0P1P2() is called by SetupClock() to compute
+	// the parameters for the connection.
+	// In comparison, when the chip is running in PCON mode, macOS recognizes the HDMI port as
+	// a DisplayPort port. As a result, this method is never called by SetupClock().
+	//
+	// This fix is left here as an emergency fallback and for reference purposes,
+	// and is compatible for graphics on Skylake, Kaby Lake and Coffee Lake platforms.
+	// Note that it is still capable of finding appropriate dividers for a 1080p HDMI connection
+	// and is more robust than the original implementation.
+	//
+	// For those who want to have "limited" 2K/4K experience (i.e. 2K@59Hz or 4K@30Hz) with their
+	// HDMI 1.4 port, you might find this fix helpful.
+	//
+	// For those who have a laptop or PC with HDMI 2.0 routed to IGPU and have HDMI output issues,
+	// it is still recommended to enable the LSPCON driver support to have full HDMI 2.0 experience.
+	//
+	// - FireWolf
+	// - 2019.06
+	//
+	
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Called with pixel clock = %d Hz.", pixelClock);
+	
+	/// All possible dividers
+	static constexpr uint32_t dividers[] = {
+		// Even dividers
+		4,  6,  8, 10, 12, 14, 16, 18, 20,
+		24, 28, 30, 32, 36, 40, 42, 44, 48,
+		52, 54, 56, 60, 64, 66, 68, 70, 72,
+		76, 78, 80, 84, 88, 90, 92, 96, 98,
+		
+		// Odd dividers
+		3, 5, 7, 9, 15, 21, 35
+	};
+	
+	/// All possible central frequency values
+	static constexpr uint64_t centralFrequencies[3] = {8400000000ULL, 9000000000ULL, 9600000000ULL};
+	
+	// Calculate the AFE clock
+	uint64_t afeClock = pixelClock * 5;
+	
+	// Prepare the context for probing P0, P1 and P2
+	struct ProbeContext context;
+	bzero(&context, sizeof(struct ProbeContext));
+	
+	// Apple chooses 400 as the initial minimum deviation
+	// However 400 is too small for a pixel clock like 533.25 MHz (HDMI 2.0 4K @ 60Hz)
+	// Raise the value to UInt64 MAX
+	// It's OK because the deviation is still bound by MAX_POS_DEV and MAX_NEG_DEV.
+	context.minDeviation = UINT64_MAX;
+	
+	for (int dindex = 0; dindex < sizeof(dividers) / sizeof(int); dindex++) {
```

Comment:
Here and in a loop below I am pretty sure it must be not be int (but size_t), and must at least be arrsize(dividers). Most likely, however, you would want to use range-based for once again:

for (auto &divider : dividers)
for (auto &centra :  centralFrequencies)

Reply:
Thanks for mentioning that. Since those arrays are immutable, I don't need the reference to the element in the array, so I think I could use the value directly.

```
for (auto divider : dividers) { ... }
```